foundations/src/telemetry/log/internal.rs

```diff
@@ -2,93 +2,59 @@ use super::init::LogHarness;
-// the Logger.
-//
-// Doing it this way means we lose the enforced atomicity between updating the
-// Logger and updating the generation number. In general, we try to only read
-// or write the generation number while we hold the appropriate lock on the
-// Logger. It is not mission-critical that we do this. The generation number
-// is just a guard rail to prevent us from going wild. Even if there is some
-// race condition and we're off by one or two, it won't affect its ability to
-// do its job.
-//
-// NOTE: we intentionally use a lock without poisoning here to not
-// panic the threads if they just share telemetry with failed thread.
-pub(crate) type SharedLog = Arc<LogContext>;
-
-#[derive(Debug)]
-pub struct LogContext {
-    // The logger itself. This is the most important part of this struct.
-    pub logger: Arc<parking_lot::RwLock<Logger>>,
-
-    // Generation number, incremented every time the logger is replaced with
-    // a child of itself. Accuracy is not critical as this is only used as a
-    // safety check. Nevertheless, when possible, please be holding the
-    // logger's lock when you read or write the generation number.
-    pub(crate) generation: AtomicU32,
+    // Nesting level, incremented every time the logger is replaced with a child of itself.
+    // Accuracy is not critical as this is only used as a safety check.
+    pub(crate) nesting_level: u32,
 }
 
-impl LogContext {
-    // Create a new LogContext based on a fresh logger. The generation number
-    // is initialized to zero.
+impl LoggerWithNestingTracking {
+    pub const MAX_LOG_NESTING: u32 = 1000;
+    pub const EXCEEDED_MAX_LOG_NESTING_ERROR: &'static str = "foundations: maximum logger nesting exceeded (are add_fields! or set_verbosity being called in a loop?)";
+
+    // Create a new LoggerWithNestingTracking based on a fresh logger. The nesting level is
+    // initialized to zero.
     pub(crate) fn new(logger: Logger) -> Self {
         Self {
-            logger: Arc::new(parking_lot::RwLock::new(logger)),
-            generation: AtomicU32::new(0),
+            inner: logger,
+            nesting_level: 0,
         }
     }
 
-    // Update this LogContext's logger object, and, atomically, increment
-    // the generation number. This mutates 'self' using interior mutability,
-    // even though it only takes a shared reference to self.
-    pub(crate) fn update<F>(&self, f: F)
+    // Update this LoggerWithNestingTracking's logger object, and increment the nesting level.
+    pub(crate) fn update<F>(&mut self, f: F)
     where
         F: FnOnce(&Logger) -> Logger,
     {
-        let mut logger_lock = self.logger.write();
-        let generation = self.generation.fetch_add(1, Ordering::SeqCst) + 1;
-        let logger = f(logger_lock.deref());
-        *logger_lock = logger;
-
-        if generation >= MAX_LOG_GENERATION {
-            panic!("{}", EXCEEDED_MAX_LOG_GENERATION_ERROR);
+        let logger = f(&self.inner);
```

Comment:
Having  a function named update is a bit vague and susceptible to misuse - updating inner log not necessary means that the nesting level will increase. Also, it has a bit unconventional argument for updating that would usually imply a mut reference to a property that supposed to be updated. I can't really come up with a good name for this function, it's really something like `replace_inner_with_new_log_with_nested_kv` and it's too verbose. It's a good sign that we probably should have the function at all - let's just update the `inner` field on a callsite and then introduce `inc_kv_nesting_level` fn that will increment nesting and do the assertion.

Reply:
Having all "replace logger with a child of itself" updates go through a single code path makes it harder to "miss one". So I felt comfortable with the unit tests only testing add_fields!, because I trusted that if add_fields! enforced the limit, so would set_verbosity().

Now that we don't have a single update() method anymore, I can't implicitly trust that anymore. But I don't want to bloat two unit tests into four, either.

So what I'm doing is I'm having each of the unit tests do approximately MAX/2 add_fields!, and approximately MAX/2 set_verbosity(). (One unit test does slightly under, the other unit test does slightly over)

I think this is sufficient, because if we missed the check in one of the two places, then the unit test that does 501 add_fields! and 501 set_verbosity() would only increment the counter 501 times, not 1002 times. So it wouldn't panic, and the test case would fail.

So I think this is good.